components/pkg-export-docker/src/graph.rs

```diff
@@ -0,0 +1,214 @@
+use crate::{build::BasePkgIdents,
+            error::Result};
+use habitat_common::package_graph::PackageGraph;
+use habitat_core::package::PackageIdent;
+use linked_hash_map::LinkedHashMap;
+use std::path::Path;
+
+pub struct Graph {
+    g:    PackageGraph,
+    base: BasePkgIdents,
+    user: Vec<PackageIdent>,
+}
+
+impl Graph {
+    pub fn from_packages(base: BasePkgIdents,
+                         user: Vec<PackageIdent>,
+                         rootfs: &Path)
+                         -> Result<Graph> {
+        let g = PackageGraph::from_root_path(rootfs)?;
+        Ok(Graph { g, base, user })
+    }
+
+    /// Helper function to create a Vec of our base idents in a
+    /// sensible order (roughly in order of package volatility).
+    ///
+    /// The idea is that we'll add packages to the image in this
+    /// order, one layer at a time, in order to try and maximize
+    /// layer caching.
+    fn idents_from_base(&self) -> Vec<PackageIdent> {
+        let mut idents = vec![];
+        if let Some(ref busybox) = self.base.busybox {
+            idents.push(busybox.clone());
+        }
+        idents.push(self.base.launcher.clone());
+        idents.push(self.base.hab.clone());
+        idents.push(self.base.sup.clone());
+        idents.push(self.base.cacerts.clone());
+        idents
+    }
+
+    /// Return the list of packages to install in the image in
+    /// dependency order.
+    ///
+    /// Note that this is essentially a consistent union of the
+    /// reverse topological sorts of all the "top level" packages that
+    /// are added to a container image.
+    ///
+    /// User packages will be last. Ideally, as users are iterating on
+    /// their packages and creating images, this should mean that all
+    /// the dependencies are already available as cached layers.
+    pub fn reverse_topological_sort(&self) -> Vec<PackageIdent> {
```

Comment:
Minor - Can the result of this calculation ever change after the graph has been created? Could we do the calculation as part of `from_packages` and just store the `Vec<PackageIdent>`?

Reply:
No, it wouldn't change, so we could technically store it at creation time.

With the current usage, it's basically a "six of one, half-a-dozen of the other" kind of situation, I think.